rohmu/object_storage/google.py

```diff
@@ -703,3 +723,98 @@ def _read_bytes(self, length: int, *, initial_data: Optional[bytes] = None) -> b
             return read_results[0]
         else:
             return b"".join(read_results)
+
+
+class MediaDownloadProtocol(Protocol):
+    def next_chunk(self) -> tuple[MediaDownloadProgress, bool]:
+        ...
+
+
+class MediaIoBaseDownloadWithByteRange:
+    """This class is mostly a copy of the googleapiclient's MediaIOBaseDownload class,
+    but with the addition of the support for fetching a specific byte_range.
+
+    """
+
+    def __init__(
+        self,
+        fd: BinaryIO,
+        request: HttpRequest,
+        chunksize: int = DOWNLOAD_CHUNK_SIZE,
+        *,
+        byte_range: tuple[int, int],
+    ) -> None:
+        """Constructor.
+
+        Args:
+          fd: io.Base or file object, The stream in which to write the downloaded
+            bytes.
+          request: googleapiclient.http.HttpRequest, the media request to perform in
+            chunks.
+          chunksize: int, File will be downloaded in chunks of this many bytes.
+          byte_range: Optional[tuple[int, int]], The byterange to fetch
+        """
+        self._fd = fd
+        self._http = request.http
+        self._uri = request.uri
+        self._chunksize = chunksize
+        self._start_range, self._end_range = byte_range
+        self._progress = 0
+        self._range_size = byte_range[1] - byte_range[0] + 1
+        self._done = False
+
+        self._headers = {}
+        for k, v in request.headers.items():
+            # allow users to supply custom headers by setting them on the request
+            # but strip out the ones that are set by default on requests generated by
+            # API methods like Drive's files().get(fileId=...)
+            if k.lower() not in ("accept", "accept-encoding", "user-agent"):
+                self._headers[k] = v
+
+    def next_chunk(self) -> tuple[MediaDownloadProgress, bool]:
+        """Get the next chunk of the download.
+
+        Returns:
+          (status, done): The value of done will be True when the media has been fully
+             downloaded or the total size of the media is unknown.
+
+        Raises:
+          googleapiclient.errors.HttpError if the response was not a 2xx (or a 416 is received and the file is empty)
+          httplib2.HttpLib2Error if a transport error has occurred.
+        """
+        headers = self._headers.copy()
+        chunk_start = self._progress + self._start_range
```

Comment:
```suggestion
        chunk_start = max(self._progress, self._start_range)
```
chunk_start should start from `self._start_range` only for the first chunk, then its always `self._position`

Reply:
> If you think about multiple calls to this functions, the _start_range will always be the same, while _progress increases. What you suggest would end up duplicating some bytes, because it will first return _start_range until _progress is bigger and then return that instead.

I might be mis-undestanding, but `self._position` gets advanced after every chunk is downloaded (`self._progress += len(content)`) . then adding `self._start_range` (to that, on next call to `next_chunk`) would be wrong ?

Reply:
I have renamed the variables to make the meaning clearer. There is no `_position` only `_num_bytes_downloaded`, which at the beginning is 0. When you want to download a chunk of the response you add `_num_bytes_downloaded` to the `_start_position` of the range to get the current position you are at where the next chunk starts, so `chunk_start = self._num_bytes_downloaded + self._start_position`.

Does this make sense to you?

Reply:
this was making it weird.
```
# pseudo code
chunk_start = self._progress + self._start_position
...
download(start_at=chunk_start)
..
self._progress += len(content))
```
but the latest change has fixed it.